redis/src/main/scala/zio/redis/Interpreter.scala

```diff
@@ -1,135 +1,236 @@
 
 import zio._
-import zio.blocking._
+import zio.stream.Stream
 
 trait Interpreter {
+
+  import Interpreter._
+
   type RedisExecutor = Has[RedisExecutor.Service]
 
   object RedisExecutor {
 
-    val ResponseBufferSize = 1024
+    val defaultPort = 6379
 
     trait Service {
-      def execute(command: Chunk[String]): IO[RedisError, String]
+
+      def execute(command: Chunk[RespValue.BulkString]): IO[RedisError, RespValue]
+
+    }
+
+    /**
+     * This is to make better use of pipelining by writing requests to the socket in batches.
+     * Ideally this*(average RESP serialized message size) should equal the buffer size used by `ByteStream`,
+     * but we can only guesstimate.
+     */
+    private val requestQueueSize = 20
+
+    private final class Live(
+      reqQueue: Queue[Request],
+      resQueue: Queue[Promise[RedisError, RespValue]],
+      byteStream: Managed[IOException, ByteStream.ReadWriteBytes]
+    ) extends Service {
+
+      override def execute(command: Chunk[RespValue.BulkString]): IO[RedisError, RespValue] =
+        Promise
+          .make[RedisError, RespValue]
+          .flatMap(promise => reqQueue.offer(Request(command, promise)) *> promise.await)
+
+      private def sendNext(out: Chunk[Byte] => IO[IOException, Unit]): IO[RedisError, Unit] =
+        reqQueue.takeBetween(1, Int.MaxValue).flatMap { reqs =>
+          val bytes = Chunk.fromIterable(reqs).flatMap(req => RespValue.Array(req.command).serialize)
+          out(bytes)
+            .mapError(RedisError.IOError)
+            .tapError(e => IO.foreach(reqs)(_.promise.fail(e)))
+            .zipLeft(IO.foreach(reqs)(req => resQueue.offer(req.promise)))
+        }
+
+      private def runReceive(inStream: Stream[IOException, Byte]): IO[RedisError, Unit] =
+        inStream
+          .mapError(RedisError.IOError)
+          .transduce(RespValue.deserialize.toTransducer)
+          .foreach(response => resQueue.take.flatMap(_.succeed(response)))
+
+      /**
+       * Opens a connection to the server and launches send and receive operations.
+       * All failures are retried by opening a new connection.
+       * Only exits by interruption or defect.
+       */
+      def run: IO[RedisError, Unit] =
+        byteStream
+          .mapError(RedisError.IOError)
+          .use { rwBytes =>
+            sendNext(rwBytes.write).forever race runReceive(rwBytes.read)
+          }
+          .tapError { e =>
+            IO.effectTotal(println(s"Reconnecting due to error: $e")) *>
+              resQueue.takeAll.flatMap(IO.foreach(_)(_.fail(e)))
+          }
+          .retryWhile(Function.const(true))
+          .tapCause(c => IO.effectTotal(Console.err.println(s"Executor exiting: $c")))
+
     }
 
-    def live(host: String, port: Int): ZLayer[Blocking, IOException, RedisExecutor] =
+    def live: ZLayer[ByteStream, RedisError.IOError, RedisExecutor] =
```

Comment:
I wouldn't leak `ByteStream`.

Reply:
It would be useful to expose it if we want the option to use something other than a TCP socket for the transport. But I admit I can't think of any situation where that's useful…

Reply:
Made it private 👍